crates/agglayer-contracts/src/lib.rs

```diff
@@ -126,10 +131,65 @@ where
     }
 
     async fn get_l1_info_root(&self, l1_leaf_count: u32) -> Result<[u8; 32], ()> {
-        self.l1_info_tree
-            .l_1_info_root_map(l1_leaf_count)
+        // Get `UpdateL1InfoTreeV2` event for the given leaf count from the latest block
+        let filter = Filter::new()
+            .address(self.l1_info_tree.address())
+            .event("UpdateL1InfoTreeV2(bytes32,uint32,uint256,uint64)")
+            .topic1(U256::from_big_endian(&l1_leaf_count.to_be_bytes()))
+            .from_block(BlockNumber::Earliest);
+
+        let events = self
+            .l1_info_tree
+            .client()
+            .get_logs(&filter)
             .await
-            .map_err(|_| ())
+            .map_err(|_| ())?;
+
+        // Extract event details
+        let (l1_info_root, event_block_number) = events
+            .first()
+            .and_then(|log| {
+                match PolygonZkEVMGlobalExitRootV2Events::decode_log(&log.clone().into()).ok()? {
+                    PolygonZkEVMGlobalExitRootV2Events::UpdateL1InfoTreeV2Filter(event) => {
+                        Some((event.current_l1_info_root, log.block_number?))
+                    }
+                    _ => None,
+                }
+            })
+            .ok_or(())?;
+
+        debug!(
+            "Retrieved UpdateL1InfoTreeV2 event from block {}. L1 info tree leaf count: {}, root: \
+             {:?}",
+            event_block_number, l1_leaf_count, l1_info_root
+        );
+
+        // Await for the related block to be finalized
+        // NOTE: Cannot use block subscription because the provider is not websocket
+        {
+            let mut tick = tokio::time::interval(tokio::time::Duration::from_secs(10));
+            let mut finalized_block_number: U64 = 0.into();
+            while finalized_block_number < event_block_number {
```

Comment:
Should there be a maximum number of attempts too?

Reply:
I added a timeout based on the time for finality on ethereum